### ISSUE TITLE
IE support: "'Promise' is undefined", bundles fail to load

### ISSUE THREAD

**Issue by edm00se** [parcel-bundler/parcel#2364]
# 🐛 bug report

Loading the dev server output in IE11 keeps bundles from loading. It's a hard stop that prevents the base page from even attempting to continue.

Note: This is from what's generated by parcel, not my app source code or my dependencies. See [this comment](https://github.com/parcel-bundler/parcel/issues/2364#issuecomment-445807464) below.

## 🎛 Configuration ([.babelrc](https://github.com/edm00se/parcel-ie11-issue-demo/blob/master/.babelrc), [package.json](https://github.com/edm00se/parcel-ie11-issue-demo/blob/master/package.json), [cli command](https://github.com/edm00se/parcel-ie11-issue-demo/blob/507761c60b4439d2bcad6fb82701edae6ee3cfd0/package.json#L10))

<details><summary><code>.babelrc</code></summary><p>
The app works in FF + Chrome (used <code>--no-cache</code>) without the custom <code>.babelrc</code>.

```js
{
  "presets": [
    [
      "env",
      {
        "modules": false,
        "targets": {
          "browsers": [
            "> 1%",
            "last 2 versions",
            "ie 11",
            "not ie <= 10"
          ]
        }
      }
    ]
  ]
}
```
</p></details>

<br>

<details><summary><code>package.json</code></summary><p>
<a href="https://github.com/edm00se/parcel-ie11-issue-demo/blob/master/package.json">here's the full <code>package.json</code></a>, below is the scripts and deps only

```json
  "scripts": {
    "clean": "rimraf dist",
    "dev": "parcel src/index.html",
    "build": "run-s clean build:raw",
    "build:raw": "parcel build --public-url . src/index.html",
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "devDependencies": {
    "babel-core": "^6.26.3",
    "babel-preset-env": "^1.7.0",
    "npm-run-all": "^4.1.5",
    "parcel-bundler": "^1.10.3",
    "rimraf": "^2.6.2"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "ie 11",
    "not ie <= 10"
  ]
```
</p></details>

<br>

I start the local server with `npm run dev`, which invokes `parcel ./src/index.html`.

Note: my `package.json` includes the relevant `browserslist`, which specifies IE 11.

## 🤔 Expected Behavior

Bundles should load correctly, including in IE 11.

## 😯 Current Behavior

Bundled app stops loading in IE11 almost immediately. Cites "'Promise' is undefined" in JS console, clicking to the line number points to the `loadBundles` function, which returns `Promise.all(bundles.map(loadBundle))`.

Also used by Parcel's generated main js file, `Promise.resolve` and `fetch`. Specifying IE 11 support for targeted browser should ensure that no un-polyfilled APIs are used (Promise, fetch).

Raw console error:
```
SCRIPT5009: 'Promise' is undefined
main.1f19ae8e.js 394,27)
```

![parcel_ie_issues](https://user-images.githubusercontent.com/622118/50187159-d7544780-02e2-11e9-9341-617b39015433.gif)

## 💁 Possible Solution

From the other linked "ie support" descriptions in [the changelog](https://github.com/parcel-bundler/parcel/blob/master/CHANGELOG.md) and the corresponding culprits, it looks like this is has to do fundamentally with the un-polyfilled `Promise` and `global` objects, such as `fetch` referenced when attempting to load necessary bundles. The `loadBundles` fires before my app logic, so no polyfilling on my end will affect the outcome. By performing a build of my repro [with manually added polyfills](https://github.com/parcel-bundler/parcel/issues/2364#issuecomment-445808271) into the `head` tag for `Promise` and `fetch`, this worked; but this isn't something I should have to do.

## 🔦 Context

For my work apps am required to support IE11. I believe this should be achievable w/ parcel, but ran into some trouble. The app works great in Chrome and Firefox. The issues I'm experiencing deal with Parcel's direct use of `Promise` and `fetch` directly in the initialization of bundles from the built `main.<hash>.js`.

## 💻 Code Sample

I have an example repo:

https://github.com/edm00se/parcel-ie11-issue-demo

Steps to reproduce:

- `git clone https://github.com/edm00se/parcel-ie11-issue-demo.git`
- `cd parcel-ie11-issue-demo`
- `npm install`
- `npm run dev`
- open `http://localhost:1234` in IE 11
- have a look at the js console, since the screen is blank (screen shots above)

## 🌍 Your Environment

| Software      | Version(s) |
| ------------- | ---------- |
| Parcel          |  1.10.3 |
| Node           | 10.13.0 |
| npm/Yarn    |  6.4.1 |
| Operating System | Windows 10 |

## Thank You! 

I love parcel, it's a great and performant bundler that's enjoyable to use. When I set out on this crazy path, I wasn't expecting to get as far as I have, and I am 99.8% there!

\*note: I've updated this description to point to a far more simple reproducible repo

**Comment by DeMoorJasper** [comment]
See https://github.com/parcel-bundler/parcel/pull/2073#issuecomment-427610992

**Comment by edm00se** [comment]
@DeMoorJasper If you check my `package.json`, it includes a browserslist definition which explicitly includes “ie 11”. With that said, the generated output still attempts to use Promise.all to load the generated bundles.

**Comment by masterwaffle** [comment]
I also hit this very problem today - seems an issue with the pre-app serving code not the app code transpiling/polyfilling itself. 

I was blown away 💨  by Parcel and how it completely simplified the build over webpack, but we pretty much exclusively target IE at work 🤦🏻‍♂️.

**Comment by DeMoorJasper** [comment]
Like stated in the comment, every npm package you use that isn't commonjs should also have a browserlist or engines field in pkg.json to indicate what syntax the code is using.
Otherwise parcel would need to assume that all modules need to be compiled if a target isn't defined, which isn't very good for performance.

**Comment by masterwaffle** [comment]
I’m confused 🤷🏻‍♂️-> isn’t the issue here that Parcel 📦 is trying to unpolyfilled Promise to load its bundles?

I understand running babel over modules which are not compiled down to es5, but isn’t the issue here that Parcel 📦 itself is trying to use the undefined Promise object?

Here’s what I did to test:
`parcel build src/index.html`
Tested in IE 11 =>produces error ❌
Added [promise polyfill](https://github.com/stefanpenner/es6-promise) manually to dist/index.html (and [fetch] (https://github.com/github/fetch/)
Works ✅ 

To me, that suggests that Promise is being used by Parcel 📦 before its possible to polyfill it (as opposed to any of the dependencies actually causing the issue).

**Comment by thefoxie** [comment]
@DeMoorJasper 
I'm trying to use it on local (no npm package) scripts, imported via ES6 import.
The event() constructor `new Event("event_name");` was easy to workaround, but I don't want to write workaround for every unsupported feature.

**Comment by edm00se** [comment]
@masterwaffle That’s what I’m seeing as well.

**Comment by edm00se** [comment]
I threw together a dead simple reproducible app, purely vanilla (no frameworks), but with configured babel which should match up with parcel's requirements from the app source perspective. This shows the same behavior I described, with less surrounding clutter. Specifically, with more than one bundle, Parcel itself relies directly on `Promise` and `fetch`.

source: https://github.com/edm00se/parcel-ie11-issue-demo
deployed: https://edm00se.codes/parcel-ie11-issue-demo/

![image](https://user-images.githubusercontent.com/622118/50017429-a34eee80-ff91-11e8-8408-895de1643ec9.png)

**Comment by edm00se** [comment]
I've updated the description of this to reflect the bare bones repro as outlined in my last comment.

**Comment by edm00se** [comment]
> Like stated in the comment, every npm package you use that isn't commonjs should also have a browserlist or engines field in pkg.json to indicate what syntax the code is using.
> Otherwise parcel would need to assume that all modules need to be compiled if a target isn't defined, which isn't very good for performance.

@DeMoorJasper I have [a repro](https://github.com/parcel-bundler/parcel/issues/2364#issuecomment-447391445) that uses no additional npm packages, beyond vanilla js and requires an html partial. There is nothing happening inside the app that uses anything over es5, the browser target should be down to "ie 11", yet parcel is directly generating direct uses to un-polyfilled `Promise` and `fetch` APIs. I have multiple bundles since I'm requiring in an `html`file, which causes this. Is there a better way of being able to require the `html` content and ensure compatibility down to es5/IE11?

**Comment by edm00se** [comment]
For any interested in this issue, I threw together a parcel plugin to add high level polyfills for both `Promise` and `fetch` to keep parcel's bundle loader working in IE(11).

repo: https://github.com/edm00se/parcel-plugin-goodie-bag
use: `npm i -S parcel-plugin-goodie-bag` (auto loads into `index.html` on build/dev)

**Comment by riverross** [comment]
TY @edm00se your plugin worked perfectly for me!

**Comment by edm00se** [comment]
I'm glad to hear it @riverross.

**Comment by edm00se** [comment]
In case it's not clear from above, part of the reason I had issues was due to:

- needing to support Internet Explorer
  - I hate it
  - [Microsoft doesn't even like it](https://techcommunity.microsoft.com/t5/Windows-IT-Pro-Blog/The-perils-of-using-Internet-Explorer-as-your-default-browser/ba-p/331732)
  - some of us poor saps are stuck supporting it due to contractual promises
- `babel-polyfill`, which is the v6 babel package, doesn't polyfill either `Promise` or `fetch
  - this appears to different for at least `Promise` with the v7 package, [@babel/polyfill](https://babeljs.io/docs/en/babel-polyfill)
  - `fetch` polyfill will still be needed
  - due to timing, pulling in a `Promise` polyfill may still be needed
- the app configuration of `babel-polyfill` fires _after_ when Parcel will need it for a multi bundle scenario
  - requiring in html partials seem to cause that scenario

I _wish_ there was an easier way of doing this, making Parcel more intelligent regarding browser support detection and doing this so I wouldn't have to. That said, it's behaving with what it was given as far as configuration. With any luck that might change in the future, until then we can work with the situation. 🤞

**Comment by koreanlike** [comment]
Thank you for your comment edm!!
your comment was very helpful:)

**Comment by tomasdev** [comment]
What's the point on even having browserslist if it won't work as expected.

**Comment by edm00se** [comment]
@tomasdev as a consumer of dev tooling, I agree. 

I think what is central here is that parcel, at least v1's approach, seems to try to try to provide default enabled and configurable "other tools". My preference would look something along the lines of detecting the `browserslist` definition and determining whether to need to polyfill `Promise` and `fetch` early in its generated code, providing a virtually seamless experience.

**Comment by edm00se** [comment]
Further activity.

**Comment by arminrosu** [comment]
@edm00se I have the same issue, only worse. In a Typescript project, we need to support IE11, which means a lot more than `Promise` and `fetch`.

.babelrc
```json
{
  "presets": [
    "@babel/preset-typescript",
    "@babel/preset-react",
    [
      "@babel/preset-env",
      {
        "debug": true,
        "corejs": {
          "version": 3
        },
        "useBuiltIns": "entry"
      }
    ]
  ],
  "plugins": [
    "@babel/plugin-transform-runtime"
  ]
}
```

.browserlist
```
last 2 versions
IE 11
not dead
```

src/index.ts (main entry point)
```ts
// IE11 polyfills >
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import 'whatwg-fetch';

// ...
```

This worked, but now I added a new dependency and that one doesn't get transpiled. There's an issue already open: #1655

The temporary fix for that was a [postinstall script](https://github.com/parcel-bundler/parcel/issues/1655#issuecomment-568175592) that modifies `node_modules/**/package.json`. 😔

Hope this helps someone else too.

**Comment by gregg-cbs** [comment]
Its 2021 and I still cant figure out how to get Parcel to work on IE11.

Zero config my balls 😂

**Comment by edm00se** [comment]
@gregg-cbs [Microsoft is doing almost everything it can to not support or be shackled to IE](https://techcommunity.microsoft.com/t5/Windows-IT-Pro-Blog/The-perils-of-using-Internet-Explorer-as-your-default-browser/ba-p/331732). It is ancient by all modern web browser standards and hasn’t received any feature updates since 2016, so I suggest that if your expectation is that it should be supported in 2021, perhaps you should revisit your expectations.

If you’re still set on making it work with IE, please note that this issue was created for Parcel v1, v2’s approach I wouldn’t expect to work with IE out of the box as Parcel is modern tooling and IE is anything but modern. I also suggest checking out the following:
- I wrote [a plug-in for parcel v1 that bundles two polyfills](https://github.com/edm00se/parcel-plugin-goodie-bag), a team I was on made use of this to get our contracted IE support met
- try out [parcel v2, which supports legacy bundles](https://github.com/edm00se/parcel-plugin-goodie-bag/issues/18#issuecomment-803409442) and makes it quite easy to eliminate the need for a plug-in like the one I just mentioned
- if you’re set on being salty about demanding support for an ancient browser at this point of IE’s horrible death spiral, try something else

Closing this issue as there’s nothing further to be gained here. There are ways of making this work for v1, v2 makes it easy even though it’s not what should be a common use case now, and with the v2 RC out the API is mostly settled and the full release is on the horizon.

**Comment by gregg-cbs** [comment]
Trust me, I tell anyone I work for that I dont support or code for IE. But sometimes a client wants what they want and it angers me as much as the next person.

This is the quote from parcels home page:
"From modern and legacy browser support, to zero config JSX and TypeScript compilation, Parcel makes it easy to build for any target – or many!"

See the webpage told me that I could expect legacy browser support with zero config. I didnt assume you would support an ancient browser, i didnt ask you to do that. Your website told me thats what I could expect. I spent a lot of time trying to implement parcel and I appologise if I came off salty but it was a time consuming exercise.

In the end I told my client they need to let go of IE.
